### PR TITLE
add pd client & send keyAddress in Get()

### PR DESCRIPTION
**Pull request by morefreeze**
Add KeyAddress in Prewrite&Commit will be complex, move them into next PR.
PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
we should remove tso, now only use pd

**Comment by morefreeze** [comment]
Yep, I will remove it after migrate pd successfully.

File: kv.go [hunk 1]
```diff
@@ -54,46 +54,139 @@ const (
...
+	// search in it firstly.
+	// FIXME: If region cache update infrequently, [start, end) can be stored in
+	// an array then use binary search the key.
+	regionCache map[uint64]*raftRegion
 }
 
-func (s *tikvStore) getClient() *Client {
-	return s.client
+func (s *tikvStore) getClient(nodeID uint64) (*Client, error) {
+	client, ok := s.clients[nodeID]
```
Comment (disksing) [inline kv.go]: clients and regionCache will be accessed by concurrent transactions, should be guarded by Mutex. An RWMutex would be suitable.

Comment (disksing) [inline kv.go]: You may refer https://github.com/pingcap/pd/blob/master/server/cluster.go#L114 for an example.

Comment (morefreeze) [inline kv.go]: Yes, I will review threadsafe.

File: kv.go [hunk 2]
```diff
@@ -54,46 +54,139 @@ const (
...
 	oracle    oracle.Oracle
-	client    *Client
+	// nodeID => *Client
+	clients  map[uint64]*Client
+	pdClient *pdClient.Client
+	// regionCache store some regions that had been requested. Get key info will
+	// search in it firstly.
+	// FIXME: If region cache update infrequently, [start, end) can be stored in
+	// an array then use binary search the key.
+	regionCache map[uint64]*raftRegion
```
Comment (siddontang) [inline kv.go]: you can use the encoded region end key to search the region. so what you need is to find the end key > given key.
You can see https://github.com/pingcap/pd/blob/master/server/cluster.go#L168 for how to encode the search key.

File: kv.go [hunk 3]
```diff
@@ -54,46 +54,139 @@ const (
...
+		s.clients[nodeID] = NewClient(meta.GetAddress())
+		client = s.clients[nodeID]
+	}
+	return client, nil
+}
+
+// getKeyAddress return k where in the region and peerLeader, if region is new,
+// pick first peer as leader to return.
+func (s *tikvStore) getKeyAddress(k []byte) (*raftRegion, *Client, error) {
+	for _, r := range s.regionCache {
```
Comment (siddontang) [inline kv.go]: no thread lock?

Comment (morefreeze) [inline kv.go]: Will add it.

**Comment by morefreeze** [comment]
PTAL

**Comment by ngaut** [comment]
PTAL @disksing

**Comment by disksing** [comment]
LGTM

**Pull request merged**